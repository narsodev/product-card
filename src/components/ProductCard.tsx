import React from 'react'

import { createContext, CSSProperties } from 'react'
import useProduct from '../hooks/useProduct'
import { InitialValues, onChangeArgs, Product, ProductCardHandlers } from '../interfaces/interfaces'

import { ProductContextProps } from '../interfaces/interfaces'

import styles from '../styles/styles.module.css'

export const ProductContext = createContext({} as ProductContextProps)
const { Provider } = ProductContext

export interface Props {
  // children?: ReactElement | ReactElement[],
  className?: string,
  initialValues?: InitialValues,
  product: Product,
  style?: CSSProperties,
  value?: number,
  children: (args: ProductCardHandlers) => JSX.Element,
  onChange?: (args: onChangeArgs) => void
}


const ProductCard = ({ product, children, className = '', style, onChange, value, initialValues }: Props) => {
  const { counter, isMaxCountReached, maxCount, increaseBy, reset } = useProduct({ product, onChange, value, initialValues })

  return (
    <Provider value={{
      counter,
      isMaxCountReached,
      product,
      increaseBy
    }}>
      <div
        className={`${styles.productCard} ${className}`}
        style={style}
      >
        {children({
          count: counter,
          isMaxCountReached,
          maxCount,
          product,
          increaseBy,
          reset
        })}
      </div>
    </Provider>
  )
}

export default ProductCard
