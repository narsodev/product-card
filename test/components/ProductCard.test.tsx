import React from 'react'
import { create, act } from 'react-test-renderer'
import ProductCard from '../../src/components/ProductCard'
import { product1 } from '../data/products'

describe('Product Card', () => {
  it('renders children', () => {
    const wrapper = create(
      <ProductCard product={product1}>
        {() => <h1>Product Card</h1>}
      </ProductCard>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('increases counter', () => {
    const wrapper = create(
      <ProductCard product={product1}>
        {({ count, increaseBy }) => (
          <>
            <h1>Product Card</h1>
            <span>{count}</span>
            <button onClick={() => increaseBy(1)}>+</button>
          </>
        )}
      </ProductCard>
    )

    expect(wrapper).toMatchSnapshot()

    const button = wrapper.root.findByType('button')
    act(() => {
      button.props.onClick()
    })

    const count = wrapper.root.findByType('span')
    expect(count.children).toEqual(['1'])
  })
})
