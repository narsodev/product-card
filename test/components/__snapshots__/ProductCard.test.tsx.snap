// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Card increases counter 1`] = `
<div
  className="productCard "
>
  <h1>
    Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    +
  </button>
</div>
`;

exports[`Product Card renders children 1`] = `
<div
  className="productCard "
>
  <h1>
    Product Card
  </h1>
</div>
`;
